File: spring/src/main/java/com/example/alten/product/controller/ProductController.java
package com.example.alten.product.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import com.example.alten.product.business.ProductBusiness;
import com.example.alten.product.dto.CreateProductDto;
import com.example.alten.product.dto.ProductResponseDto;
import com.example.alten.product.dto.UpdateProductDto;

import jakarta.validation.Valid;

import java.util.List;

@RestController
@RequestMapping("/products")
public class ProductController {

    private final ProductBusiness productBusiness;

    @Autowired
    public ProductController(ProductBusiness productBusiness) {
        this.productBusiness = productBusiness;
    }

    /**
     * Entry point to create a new product
     * 
     * @param createProductDto Request DTO
     * @returns The newly created product
     */
    @PostMapping
    public ResponseEntity<ProductResponseDto> create(@Valid @RequestBody CreateProductDto createProductDto) {
        ProductResponseDto createdProduct = productBusiness.create(createProductDto);
        return new ResponseEntity<>(createdProduct, HttpStatus.OK);
    }

    /**
     * Entry point to list all products
     * 
     * @returns The list of all products
     */
    @GetMapping
    public ResponseEntity<List<ProductResponseDto>> findAll() {
        List<ProductResponseDto> products = productBusiness.findAll();
        return new ResponseEntity<>(products, HttpStatus.OK);
    }

    /**
     * Entry point to find one product by id
     * 
     * @param id id of the target product
     * @returns The product if found
     * @throws 404 - Not Found error if not found
     */
    @GetMapping("/{id}")
    public ResponseEntity<ProductResponseDto> findOne(@PathVariable Long id) {
        ProductResponseDto product = productBusiness.findOne(id);
        return new ResponseEntity<>(product, HttpStatus.OK);
    }

    /**
     * Entry point to update a product by id
     * 
     * @param id               target id
     * @param updateProductDto Request DTO
     * @returns The updated product if it exists
     * @throws 404 - Not Found error if the product is not found
     */
    @PatchMapping("/{id}")
    public ResponseEntity<ProductResponseDto> update(@PathVariable Long id,
            @Valid @RequestBody UpdateProductDto updateProductDto) {
        ProductResponseDto updatedProduct = productBusiness.update(id, updateProductDto);
        return new ResponseEntity<>(updatedProduct, HttpStatus.OK);
    }

    /**
     * Entry point to delete a product by id
     * 
     * @param id
     */
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> remove(@PathVariable Long id) {
        productBusiness.remove(id);
        return new ResponseEntity<>(HttpStatus.NO_CONTENT);
    }
}

File: nest/src/product/product.controller.ts
import {
    Controller,
    Get,
    Post,
    Body,
    Patch,
    Param,
    Delete,
    ParseIntPipe,
    HttpCode,
} from '@nestjs/common';
import { CreateProductQueryDto } from './dto/create-product-query.dto';
import { UpdateProductQueryDto } from './dto/update-product-query.dto';
import { ProductBusiness } from './product.business';
import { ProductResponseDto } from './dto/product-response.dto';
import { ApiTags } from '@nestjs/swagger';

@ApiTags('Products')
@Controller('products')
export class ProductController {
    constructor(private readonly productBusiness: ProductBusiness) {}

    /**
     * Entry point to create a new product
     * @param createProductDto Request DTO
     * @returns The newly created product
     */
    @Post()
    @HttpCode(200)
    create(
        @Body() createProductDto: CreateProductQueryDto,
    ): Promise<ProductResponseDto> {
        return this.productBusiness.create(createProductDto);
    }

    /**
     * Entry point to list all products
     * @returns The list of all products
     */
    @Get()
    findAll(): Promise<ProductResponseDto[]> {
        return this.productBusiness.findAll();
    }

    /**
     * Entry point to find one product by id
     * @param id id of the target product
     * @returns The product if found
     * @throws 404 - Not Found error if not found
     */
    @Get(':id')
    findOne(
        @Param('id', ParseIntPipe) id: number,
    ): Promise<ProductResponseDto> {
        return this.productBusiness.findOne(id);
    }

    /**
     * Entry point to update a product by id
     * @param id target id
     * @param updateProductDto  Request DTO
     * @returns The updated product if it exists
     * @throws 404 - Not Found error if the product is not found
     */
    @Patch(':id')
    update(
        @Param('id', ParseIntPipe) id: number,
        @Body() updateProductDto: UpdateProductQueryDto,
    ): Promise<ProductResponseDto> {
        return this.productBusiness.update(id, updateProductDto);
    }

    /**
     * Entry point to delete a product by id
     * @param id
     */
    @Delete(':id')
    remove(@Param('id', ParseIntPipe) id: number): Promise<void> {
        return this.productBusiness.remove(id);
    }
}
